decode/pngg2gini.1
." $Id: pngg2gini.1,v 1.0 2002/04/04 10:00:00 yoksas Exp $
.TH PNGG2GINI 1 "2004/09/16 10:00:00 $"
.SH NAME
pngg2gini \- Unidata LDM decoder for PNG-compressed NOAAPORT GINI imagery
.SH SYNOPSIS
.HP
.ft B
pngg2gini
.nh
\%[-f\ \fIPNGfile\fP]
\%[-v]
\%[-x]
\%[-l\ \fIlogpath\fP]
\%\fIIMGfile\fP
.hy
.ft
.SH DESCRIPTION
.LP
This program is an LDM decoder for NOAAPORT image data of type \fBGINI\fP
whose data portion (not header) has been PNG-compressed.
It reads a PNG-compressed GINI data product from standard input or from a file
and decodes that data into a NOAAPORT format GINI file.
.LP
This program can be invoked by an entry in the LDM \fBpqact.conf\fP
configuration file similar one of the following:
.sp
.nf
#
# PNG compressed NOAAPORT GINI images
#
NIMAGE	^sat/ch[0-9]/.*/(.*)/([12][0-9])([0-9][0-9])([01][0-9])([0-3][0-9]) ([0-2][0-9])([0-5][0-9])/(.*)/(.*km)/
	PIPE	-close
	pngg2gini -vl logs/ldm-mcidas.log
	/data/ldm/gempak/images/sat/\\8/\\9/\\1/\\1_\\2\\3\\4\\5_\\6\\7

# 
# PNG compressed 1 km NEXRAD Level III composite image in GINI format
#
FNEXRAD	^rad/NEXRCOMP/(...)/(...)_(........)_(....)
	PIPE	-close
	pngg2gini -vl logs/ldm-mcidas.log
	/data/ldm/gempak/nexrad/NEXRCOMP/\\1/\\2/\\2_\\3_\\4

.fi
.RE
.LP
Additional examples of LDM pqact.conf entries can be
found in the file ldm-mcidas-pqact.conf in the \fBetc/\fP subdirectory of the 
ldm-mcidas installation tree.
.fi
.RE
.LP
This program can also be invoked as a
stand-alone program, outside the LDM environment.

.SH FLAGS
.TP
.BI "-f " PNGfile
Pathname for input PNG compressed GINI file (default is stdin)
.TP
.B "-h "
Print help information, then exit the program.
.TP
.B -v
Verbose logging.
Informative messages (level LOG_INFO) are logged.
By default, only messages of severity LOG_NOTICE and greater are
logged. See the Unidata  LDM Site  Manager's  Guide  for more information on 
message logging.
.TP
.B -x
Debug logging.
Debugging messages (level LOG_DEBUG) are logged. See the Unidata  LDM Site  
Manager's  Guide  for more information on message logging.
.TP
.BI "-l " logpath
Log file pathname.
The program uses Unidata's \fBulog\fP(3) package to write error and diagnostic
messages.
The default is to write to the \fBsyslogd\fP(8) daemon.  If this option is
specified, however, then messages will be written to the file
\fIlogpath\fP unless that pathname is "-", in which case all messages will be
written to standard error.
.TP
.BI "IMGfile "
Output GINI pathname mask (required: no default).
.sp
IMGfile may contain 'replaceables':
  \\AREA - areal coverage information
  \\BAND - band information
  \\RES  - resolution information
  \\SAT  - satellite platform information
.sp
and 'strftime' style day/time specifiers that will get resolved by image
day/time information:
  %%Y    - image year       [CCYY]
  %%m    - image month      [MM]
  %%d    - image day        [DD]
  %%j    - image Julian day [JJJ]
  %%H    - image hour       [HH]
  %%M    - image minute     [MM]
  etc.

.RE
.SH NOTES
.LP
.RE
.SH "SEE ALSO"
.LP
.BR ~ldm-mcidas/etc/ldm-mcidas-pqact.conf,
.BR ldmd (1),
.BR ulog (3),
.BR syslogd (8).


RELEASE_NOTES
VERSION	COMMENTS
-------	-------------------------------------------------------------

2012    2012-01-18

        Updated version number to 2012 to reflect small updates (e.g.,
        new src/port/cfortran/cfortran.h, modified src/decode/Makefile)
        and that the package is still viable.

2008    2009-04-30

        Note: The version number is kept at 2008 since McIDAS went
        into maintenance mode with v2008.

        A variety of changes to C routines, Makefiles and test data
        files.

        Introduction of a new McIDAS lightning decoder, 'lgt2md', that
        handles both NLDN binary and USPLN ASCII data record formats.
        'lgt2md' is designed as a drop-in replacement for 'nldn2md':
        user's pattern action file entries for decoding NLDN data
        can be easily updated by replacing 'nldn2md' with 'lgt2md'.

        ROUTE.SYS (distributed as route.sys) was updated to include
        a new product, 'LP', for USPLN lightning data.

        The Makefile in the src/decode directory was reorganized to
        build the subset of the encoders/decoders contained in the package
        that are written entirely in C:

        area2png    - PNG-compresses the image portion of single-banded
                      McIDAS AREA files
        pnga2area   - uncompresses images created by 'area2png'
        zlibg2gini  - uncompresses the Zlib-compressed images contained
                      in the NOAAPort broadcast (and conveyed in the
                      IDD in the NIMAGE feed)
        pngg2gini   - uncompress NOAAPort images that have been
                      PNG-compressed by the Unidata NOAAPort ingest
                      package (archaic)
        proftomd    - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files
        lgt2md      - decodes NLDN and USPLN lightning data into
                      McIDAS MD files

        This set is built by running 'make' or 'make basic'.
        
        In addition to this set, the Fortran-based 'nldn2md' NLDN
        lightning decoder will be built by 'make full'.

        The Makefile in src/decode also provides for the (attempted)
        building of archaic McIDAS decoders:

        cdftomd     - decodes wind profiler netCDF files from NOAA/GSD
                      into McIDAS MD format files (use 'proftomd' instead)
        gunrv2      - decodes compressed McIDAS GRID files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      GRID format
        lwfile
        lwtmd2      - decodes compressed McIDAS MD files sent in the
                      old Zephyr/Alden satellite broadcast into McIDAS
                      MD format
        lwtoa3      - decodes delta-compressed McIDAS image files from
                      the old Zephyr/Alden satellite broadcast into McIDAS
                      AREA format
        nids2area   - converts Nexrad Level III images into McIDAS
                      AREA file format (best to use McIDAS IMGCOPY from
                      NEXR to AREA datasets directly as McIDAS capabilities
                      are kept up to date)
 
        In addition decoders built by 'make full', an attepmt will be
        made to build these decoders when the user runs 'make all'.
        There is _no_ guarantee that these decoders will be built
        successfully!

2004    2004-09-14

        Added O_TRUNC flag to mkdirs_open call in pnga2area.c,
        pngg2gini.c, and zlibg2gini.c.  This corrects a problem
        encountered when decoding Unidata-Wisconsin images (IDD
        feedtype UNIWISC aka MCIDAS) when using McIDAS routing table
        (ROUTE.SYS) concepts.  The problem was manifested in decoded
        image files keeping the size of the images they were
        overwriting.  So, the file size was always the "high water
        mark" of sizes for the products.

        NOTES:
        - this did not affect the sizes of files decoded without using
          McIDAS routing table concepts
        - the decoded images using McIDAS routing table concepts were
          valid/usable even though they are larger than they need to
          be

        Changed IDD datastream name from MCIDAS to UNIWISC in example
        pqact.conf files.

        Updated distribution with the versions of SATANNOT, SATBAND,
        and SCHEMA from the latest Unidata McIDAS distribution, v2004.

        Cleaned up batch.k shell script that runs McIDAS postprocess
        BATCH files.  Internal comments should now be easier to follow
        than in previous versions.

        Updated man pages for currently used decoders: area2png(1),
        nldn2md(1), pnga2area(1), pngg2gini(1), proftomd(1), and
        zlibg2gini(1).

        Removed building of decoders now considered to be obsolete:
        cdftomd(1), lwtoa3(1), gunrv2(1), lwfile(1), lwtmd2(1),
        and nids2area(1).  Source code for these decoders is still
        included in the distribution, but their use is not guaranteed
        mainly since subroutines they use from the McIDAS library
        are no longer supported.

2003    2003-07-29

        Added utililty routine zlibg2gini that uncompresses Zlib
        compressed GINI images in NIMAGE IDD datastream.

        Fixed a bug that kept McIDAS routing table, ROUTE.SYS,
        dates and times from being updated.

2002    2002-12-20

        Replaced information messages in pngg2gini.c that referred to
        pnga2area with pngg2gini.

        Corrected logic in pnga2area.c for processing a product whose
        product code is not contained in the McIDAS routing table,
        ROUTE.SYS.  Before, the routing table would get corrupted
        and the AREA file number included in the '-r pcode,anum'
        flag would not be used.  Now, the routing table will not
        be updated for unregistered products, and the default AREA
        file number will be used.
	

decode/pngg2gini.c
#define UD_FORTRAN_SOURCE

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <signal.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <limits.h>
#ifndef PATH_MAX
#  define PATH_MAX 255
#endif

#include "png.h"
#include "mkdirs_open.h"
#include "ulog.h"
#include "alarm.h"

#define WMO_HED_LEN 21
#define GINI_HED_LEN 533

/*
** Some needed function prototypes <<<<< UPC add 20080918 >>>>>
*/
void eat_stream_and_exit( int );
void readpng_row( char * );

/*
** The following are needed since this links against pngsubs.o which needs
** to have these defined.
*/

char            *annotfile=NULL;
char            *bandfile=NULL;

char             mcpathdir[PATH_MAX];                /* McIDAS MCPATH string */
char             pathname[PATH_MAX];        /* fully qualified file pathname */

int              rtecom[64];

/*
** On to declarations that are used
*/

char             pathname[PATH_MAX];        /* fully qualified file pathname */

int              READ_ERROR=0;

unsigned         alarm_timeout = 60; /* seconds; should suffice */

png_size_t       pngcount=0;
png_structp      png_ptr;
png_infop        info_ptr;
png_infop        end_info;

/*
** Function prototypes
*/

char *GiniFileName( unsigned char *imghed, char *format );
void  swbyt4_( void *, int * );

void  unidata_read_data_fn( png_structp png_p, png_bytep data, png_uint_32 length );
void  read_row_callback( png_structp pngp, png_uint_32 rownum, int pass );

/*
 * SIGALRM handler.
 */
static void
alarm_handler( int sig )
{
    assert( SIGALRM == sig );
    uerror( "pngg2gini: no data read in %u seconds, exiting", alarm_timeout );
    exit( 1 );
}

/*
** Called at exit.
** This callback routine registered by atexit().
*/

static void
cleanup()
{
    (void) closeulog();
}

/*
** Called upon receipt of signals.
** This callback routine registered in set_sigactions() .
*/

static void
signal_handler( int sig )
{
    switch( sig ) {
      case SIGHUP:
                   udebug( "SIGHUP" ) ;
                   return ;
      case SIGINT:
                   unotice( "Interrupt" ) ;
                   exit( 0 );
      case SIGTERM:
                   udebug( "SIGTERM" ) ;
                   exit( 0 );
      case SIGUSR1:
                   udebug( "SIGUSR1" ) ;
                   return;
      case SIGUSR2:
                   if ( toggleulogpri(LOG_INFO) )
                     unotice( "Going verbose" ) ;
                   else
                     unotice( "Going silent" ) ;
                   return;
      case SIGPIPE:
                   unotice( "SIGPIPE" ) ;
                   exit( 0 );
    }

    udebug( "signal_handler: unhandled signal: %d", sig );
}

static void
set_sigactions()
{
    struct sigaction sigact ;

    sigact.sa_handler   = signal_handler;
    sigemptyset(&sigact.sa_mask) ;
    sigact.sa_flags     = 0 ;

    (void) sigaction( SIGHUP,  &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGINT,  &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGTERM, &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGUSR1, &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGUSR2, &sigact, (struct sigaction *) NULL );
    (void) sigaction( SIGPIPE, &sigact, (struct sigaction *) NULL );
}

static void usage(argv0)
char *argv0;
{
#define USAGE_FMT "\
Example Usage:\n\
       %s <-vxl-> IMGfile < infile \n\
       %s <-vxl-> -f infile IMGfile\n\
Flags:\n\
    Help:\n\
    -h              Print the help, then exit the program\n\n\
    Satellite definitions:\n\
    -f infile       Pathname for input PNG compressed file (default is stdin)\n\n\
    Logging:\n\
    -l logname      Log file name (default uses local0; use '-' for stdout)\n\
    -v              Verbose logging\n\
    -x              Debug logging\n\n\
Positionals:\n\
    IMGfile         Output image pathname mask (no default)\n\n\
                    'IMGfile' may contain 'replaceables':\n\
                      \\AREA - image coverage info\n\
                      \\BAND - image band info\n\
                      \\RES  - image resolution info\n\
                      \\SAT  - satellite platform info\n\n\
                    and 'strftime' style day/time specifiers:\n\
                      %%Y    - image year       [CCYY]\n\
                      %%m    - image month      [MM]\n\
                      %%d    - image day        [DD]\n\
                      %%j    - image Julian day [JJJ]\n\
                      %%H    - image hour       [HH]\n\
                      %%M    - image minute     [MM]\n\
                      etc.\n\n\
Example Invocation:\n\
    ./pngg2gini -l - -f pngfile data/images/\\AREA/\\RES/\\BAND/\\BAND_%%Y%%m%%d.%%H%%M\n\n\
might expand to the equivalent:\n\
    ./pngg2gini -l - -f pngfile data/images/NHEM-COMP/8km/6.8um/20000511.2000\n\n"

    (void) fprintf (stderr, USAGE_FMT, argv0, argv0, argv0, argv0);
}

static char *logfname = "";

int
main( int argc, char *argv[] )
{
 
    char         *infname  = NULL;
    char         *outfname = NULL;
    char         *p        = NULL;

    char         *rowp;

    char          wmoprd[6];

    char          wmohed[WMO_HED_LEN+1];
    unsigned char imghed[GINI_HED_LEN];

    int           ch;
    int           i;
    int           nbytes;
    int           ok;

    int           bit_depth        = 8;
    int           color_type       = PNG_COLOR_TYPE_GRAY;
    int           interlace_type   = PNG_INTERLACE_NONE;
    int           compression_type = PNG_COMPRESSION_TYPE_DEFAULT;
    int           filter_type      = PNG_FILTER_TYPE_DEFAULT;

    int           logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_NOTICE)) ;
    int           logfd;

    int           fdo;

    png_uint_32   height,width;

    mode_t        mode = (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);

    long          ntot = 0;

    extern int    optind;
    extern int    opterr;
    extern char  *optarg;

    int           alarm_initialized = 0;
    Alarm         timeout_alarm;

    /*
    ** If nothing was specified on the command line, print usage and exit.
    */
    if ( argc == 1 ) {
      usage( argv[0] );
      exit( 0 );
    }

    opterr = 1;

    while ( (ch = getopt(argc, argv, "hf:vxl:")) != EOF )
      switch (ch) {
        case 'h':
        case '?':
                 usage( argv[0] );
                 exit( 0 );
                 break;
        case 'f':
                 infname = optarg;
                 break;
        case 'v':
                 logmask |= LOG_MASK(LOG_INFO);
                 break;
        case 'x':
                 logmask |= LOG_MASK(LOG_DEBUG);
                 break;
        case 'l':
                 if ( optarg[0] == '-' && optarg[1] != 0 ) {
                   fprintf( stderr, "logfile \"%s\" ??\n", optarg );
                   usage( argv[0] );
                   exit( 0 );
                 }
                 /* else */
                 logfname = optarg;
                 break;
      }

    /*
    ** Sanity checks
    */

    if ( argc - optind < 0 ) {
      usage( argv[0] );
      exit( 0 );
    }

    if ( argc != (optind + 1) ) {          /* output file name not specified */
      printf( "No output file name specified, exiting\n" );
      exit( 1 );
    }


    /*
    ** If user specified an input file with '-f' flag, redirect
    ** stdin to the file aborting on error.
    */

    if ( infname != (char *) NULL ) {
      if ( freopen( infname, "r", stdin ) == (FILE *) NULL ) {
        printf( "Can't open input file: %s\n", infname );
        exit( 1 );
      }
    }


    /*
    ** Setup syslog logging.
    */

    (void) setulogmask( logmask );

    if ( logfname == NULL || !(*logfname == '-' && logfname[1] == 0) )
      (void) fclose( stderr );

    logfd = openulog( ubasename(argv[0]), (LOG_CONS|LOG_PID), LOG_LDM,
                      logfname );

    if ( logfname == NULL || !(*logfname == '-' && logfname[1] == 0) ) {
      setbuf( fdopen(logfd, "a" ), NULL);
    }


    /*
    ** Start
    */

    unotice( "Starting Up" );

    /*
    ** Setup PNG pointers, etc.
    */

    png_ptr = png_create_read_struct( PNG_LIBPNG_VER_STRING, NULL, NULL, NULL );
    if ( !png_ptr ) exit( 2 );

    info_ptr = png_create_info_struct( png_ptr );
    end_info = png_create_info_struct( png_ptr );

    if ( !info_ptr ) {
      png_destroy_write_struct( &png_ptr, (png_infopp) NULL );
      exit( 3 );
    }

    if ( setjmp(png_ptr->jmpbuf) ) {
      png_destroy_read_struct( &png_ptr, &info_ptr, &end_info );
      exit( 4 );
    }

    png_set_read_fn( png_ptr, info_ptr, (png_rw_ptr)unidata_read_data_fn );
    png_set_read_status_fn( png_ptr, read_row_callback );


    /*
    ** Register exit handler, establish signal handlers, and set alarm handler.
    */

    (void) atexit( cleanup );

    set_sigactions();

    if ( !alarm_initialized ) {
      alarm_init( &timeout_alarm, alarm_timeout, alarm_handler );
      alarm_initialized = 1;
    }

    alarm_on( &timeout_alarm );


    /*
    ** Determine if the input file header looks like a GINI image
    */

    nbytes = GINI_HED_LEN;
    unidata_read_data_fn( png_ptr, (png_bytep) imghed, (png_uint_32) nbytes );

    if ( READ_ERROR ) {
      uerror( "GINI header read error" );
      eat_stream_and_exit( 5 );
    }

    memcpy( wmohed, imghed, WMO_HED_LEN );
    wmohed[WMO_HED_LEN] = '\0';

    if ( strncmp( (char *)imghed, "TI", 2 ) ) {
      (void) memcpy( wmoprd, imghed, 6 );
      wmoprd[5] = '\0';
      uerror( "WMO Header not supported by pngg2gini: %s", wmoprd );
      eat_stream_and_exit( 6 );
    }

    if ( imghed[21] != 1 ) {
      uerror( "Input fails Source test: 1 != %c", imghed[21] );
      eat_stream_and_exit( 7 );
    }

    if ( imghed[63] == 128 ) {              /* zero out the compression flag */
      imghed[63] = 0;
    }


    /*
    ** Create the output file name:
    **
    **   IMGfile  - file name or mask specified.  In this case, we need to
    **              expand the "replaceables" (\BAND, \RES, and \SAT) and
    **              strftime format specifications (e.g., %Y, %m, %d, etc.)
    **              in user specified filename using information from the
    **              GINI file header.
    */

    outfname = GiniFileName( imghed, argv[optind] );
    if ( outfname == (char *) NULL ) {
      uerror( "Error creating output name from pattern: %s", argv[optind] );
      eat_stream_and_exit( 8 );
    }

    uinfo( "output file pathname: %s", outfname );

    /* <<<<< UPC mod 20040914 - add file truncation >>>>> */
    fdo = mkdirs_open( outfname, O_CREAT|O_WRONLY|O_TRUNC, mode );
    if ( fdo == -1 ) {                /* make sure directories to file exist */
      uerror( "Can't open file %s", outfname );
      eat_stream_and_exit( 9 );
    }

    write( fdo, imghed, nbytes );          /* write the header w/o byte flip */


    /*
    ** Read the PNG informational header
    */

    png_read_info( png_ptr, info_ptr );
    png_get_IHDR( png_ptr, info_ptr, &width, &height, &bit_depth, &color_type,
                  &interlace_type, &compression_type, &filter_type );


    /*
    ** Allocate a buffer into which image lines are unPNGed.
    */

    rowp = (char *) malloc( width );
    if ( rowp == (char *) NULL ) {
      uerror( "Unable to malloc space for AREA row" );
      eat_stream_and_exit( 10 );
    }


    /*
    ** UnPNG an image line at a time and write it to the output file.
    */

    for ( i=0; i<height; i++ ) {
      readpng_row( rowp );
      write( fdo, rowp, width );
      ntot += width;
    }


    /*
    ** Free malloc/realloc arrays.
    */

    alarm_off( &timeout_alarm );

    free( rowp );


    /*
    ** Close the output file and exit.
    */

    close( fdo );
    fclose( stdin );

    ntot     += nbytes;                       /* all bytes written to output */
    pngcount += 16;                         /* compressed bytes + PNG header */

    unotice( "unPNG:: %8ld  %8ld  %6.4f", pngcount, ntot,
              ((float) ntot) / pngcount );

    unotice( "Exiting" );

    exit( 0 );
}
